src/rotation_types.jl

```diff
@@ -0,0 +1,192 @@
+
+
+# a default element type
+macro DefaultElType(); Float64; end  # need this to be a macro for type stability in type_methods.jl promote_type_sp functions
+
+# some trait style stuff
+numel{T}(::Type{T}) = length(fieldnames(T))    # function to get the number of elements in each parameterization
+n_params{T}(::Type{T}) = length(T.parameters)  # the number of template parameters
+
+###################################################
+# Rotation matrix (typeofalias of Mat{3,3,Float64})
+# N.B. all other parameterizations T should implement
+# a convert_rotation(T, RotMatrix) method
+###################################################
+
+@doc """
+A type alias for rotation matrices (alias of Mat{3,3,T})
+""" ->
+typealias RotMatrix{T <: Real}   Mat{3,3,T}
+#immutable RotMatrix{T <: AbstractFloat} <: FixedArray{T,2,Tuple{3,3}} # this would be better but much more effort without an implicit conversion
+#    _::Tuple{Tuple{T, T, T}, 
+#             Tuple{T, T, T},
+#             Tuple{T, T, T}}
+#end
+
+ 
+# fix weirdness, try: mat = Mat(eye(3)); @code_warntype convert(typeof(mat), mat)  
+convert{T}(::Type{RotMatrix{T}}, X::RotMatrix{T}) = X
+
+# RotMatrix(eye(3)) hangs, so fix it
+# can't use the union types because it wont replace whats in fixed size arrays
+Base.convert{U}(::Type{RotMatrix}, mat::Matrix{U}) = RotMatrix{U}(mat)
+Base.call{U}(::Type{RotMatrix}, mat::Matrix{U}) = convert(RotMatrix{U}, mat)
+
+# default parameters
+default_params{T <: RotMatrix}(::Type{T}) = (@DefaultElType(), ) 
+
+numel(::Type{RotMatrix}) = 9     # special case
+n_params(::Type{RotMatrix})  = 1 # The behaviour of length(RotMatrix.parameters) seems unstable...
+
+
+
+###################################################
+# Quaternions
+# (Quaternions are defined in Quaternions.jl)
+###################################################
+
+# add an indexing scheme
+# getindex(X::Quaternion, idx::Integer) = getfield(X, idx)
+
+# an extra constructor
+call{T}(::Type{Quaternion{T}}, a::Real, b::Real, c::Real, d::Real) = Quaternion(T(a), T(b), T(c), T(d))
+
+# define its interaction with other angle representations
+convert_rotation{T <: Real}(::Type{RotMatrix{T}}, q::Quaternion{T}) = quat_to_rot(q)
+convert_rotation{T <: Real}(::Type{Quaternion{T}}, R::RotMatrix{T}) = rot_to_quat(R)
+
+# more general
+convert_rotation{T,U}(::Type{T}, q::Quaternion{U}) = convert_rotation(T, quat_to_rot(promote_eltype(q, eltype(T))))  # go via the rotation representation
```

Comment:
Yeah, lines like 

```
convert{T}(::Type{T}, q::Quaternion) = ...
```

conflicted with everything.  I had to go with a little code gen to make specific conversions